chainstate/src/detail/tests/test_framework/framework.rs

```diff
@@ -15,15 +15,156 @@
 //
 // Author(s): A. Sinitsyn, S. Tkach
 
-use crate::detail::tests::*;
 use chainstate_storage::BlockchainStorageRead;
 use common::{
     chain::{
-        block::ConsensusData, Block, GenBlock, Genesis, OutputSpentState, Transaction, TxInput,
-        TxOutput,
+        block::{timestamp::BlockTimestamp, ConsensusData},
+        config::{Builder as ChainConfigBuilder, ChainType},
+        signature::inputsig::InputWitness,
+        Block, Destination, GenBlock, Genesis, NetUpgrades, OutPointSourceId, OutputSpentState,
+        Transaction, TxInput, TxOutput,
     },
-    primitives::{Id, Idable, H256},
+    primitives::{time, Amount, Id, Idable, H256},
 };
+use crypto::random::Rng;
+
+use crate::{
+    detail::{
+        tests::{
+            create_new_outputs, produce_test_block,
+            test_framework::{BlockTestFrameworkBuilder, ProcessBlockBuilder},
+            TestBlockInfo, ERR_CREATE_BLOCK_FAIL, ERR_CREATE_TX_FAIL,
+        },
+        BlockIndex, GenBlockIndex, TimeGetter,
+    },
+    BlockError, BlockHeight, BlockSource, Chainstate, ChainstateConfig, PropertyQueryError,
+};
+
+/// The `Chainstate` wrapper that simplifies operations and checks in the tests.
+pub struct BlockTestFramework {
+    // TODO: FIXME: Private fields?
+    pub chainstate: Chainstate,
+    // TODO: FIXME: Remove?..
+    pub block_indexes: Vec<BlockIndex>,
+}
+
+impl BlockTestFramework {
+    /// Creates a new `BlockTestFramework` instance using a builder api.
+    pub fn new() -> BlockTestFrameworkBuilder {
+        BlockTestFrameworkBuilder::new()
+    }
+
+    /// Processes a new block with the parameters specified using `ProcessBlockBuilder`.
+    pub fn process_block(&mut self) -> ProcessBlockBuilder {
+        ProcessBlockBuilder::new(self)
+    }
+
+    /// Creates and processes a given amount of blocks. Returns the if of the last produced block.
+    pub fn create_chain(
+        &mut self,
+        parent_block_id: &Id<GenBlock>,
+        count_blocks: usize,
+        rng: &mut impl Rng,
+    ) -> Result<Id<GenBlock>, BlockError> {
+        let mut prev_block = TestBlockInfo::from_id(&self.chainstate, *parent_block_id);
+
+        for _ in 0..count_blocks {
+            // TODO: FIXME:
+            let block = produce_test_block(prev_block, rng);
+            prev_block = TestBlockInfo::from_block(&block);
```

Comment:
The `TestBlockInfo` is rather hacked together to make the genesis/block split work. I don't like the fact that you always have to do this after creating a new block to get info for the next iteration. If you can think of a way of making this better or at least streamline the interface, that would be nice.

Reply:
Yes, we can have the genesis block only on the first iteration, but so far I was unable to find a solution that doesn't make the loop uglier... I'm going to try a little bit more, though.

Reply:
Some context (but not much) is here: https://github.com/mintlayer/mintlayer-core/pull/299#discussion_r927465928

Reply:
For now I'm going to add TODO notes, but I'm going to think about it.
(https://github.com/mintlayer/mintlayer-core/issues/312)

By the way, the link works for me.